### ISSUE TITLE
The Diverge 3 requires at least two layers for changes to be applied

### ISSUE THREAD

**Issue by philc** [blahlicus/arbites-family#23]
Uploading a layout with just one layer has no effect.

Steps to reproduce:

1. Download and open the default firmware for Diverge 3
2. Change a key on layer 0
3. Delete layer 1, so that there's only one layer
4. Upload

See that your change from step 2 isn't applied.

Now, add another layer and reupload. Now the changes from step 2 have been applied.

The custom layout I use only requires one layer, and so for awhile I thought the uploader program was broken because it was no uploading and applying my changes. It took me a frustrating while to stumble upon the fact that a second layer was required.

**Comment by raquo** [comment]
Yep, I've been having this problem too with Diverge 3 and Arbites Innova 2.3 5000, forgot to file a bug for this. The obvious workaround I'm sure you now know is to just have an unused second layer in your layout.

**Comment by philc** [comment]
Yep. The main reason to fix this bug is to save new users from running into this frustration.

**Comment by blahlicus** [comment]
Hey, I just tried replicating this issue on 2.3 5000 and the unreleased 2.5 0003 and failed to replicate the problem on both versions, below is my specs and what I did: (I followed the exact steps)

* OS: Win 10 running the WPF version of Arbites
* Software: Arbites 2.3 5000 and 2.5 0003
* Keyboard: I tried a few times with the following boards (Diverge 3 left side by itself, Diverge 3 both sides connected via TRRS, Felix)
* Firmware: 2.7 (The communication protocol is unchanged since 2.0 so any 2.x version should work the same)

1. Download and open the default firmware for Diverge 3
2. Change a key on layer 0
3. Delete layer 1, so that there's only one layer
4. Upload

The changes applied successfully, changed keys on `layer0` was applied and worked as intended. I was able to observe an unrelated bug though, `layer1` is retained even when `layer1` is deleted, that is, the previously loaded layout of `later1` is kept so if you press FN in this new layout, you could still access the keys from the previous `layer1`.

I think it is unlikely to be an animus issue, it is probably a division by 0 issue related to how arbites converts the layout data to serial commands and how it counts layers, but I checked the source code (at `ClLayoutContainer.GenerateCommand()`) and could not find anything.

Could one of you please send me the serial upload command outputs of this bugged behaviour to me? To do so, please start Arbites, then go to Preferences, then tick the "Display command outputs" option, then replicated the bug as per the thread OP.

**Comment by philc** [comment]
I will try to do so soon. Possibly related: I'm using a Mac.

**Comment by raquo** [comment]
@blahlicus Sorry I didn't see your reply. I have more details now that I went to change my layout and ran into this bug again.

Here are the two debug logs. upload_log_2.txt is from an upload that didn't work, and upload_log_3.txt from an upload that did work because it had an extra layer (more on this below).

[upload_log_2.txt](https://github.com/blahlicus/arbites-family/files/3338506/upload_log_2.txt)

[upload_log_3.txt](https://github.com/blahlicus/arbites-family/files/3338507/upload_log_3.txt)

The layouts of those uploads are exactly the same, except that v3 has a third layer added to it. And you can actually see this if you diff the two logs.

Now, I'll explain what didn't work and in which cases. It will take a while, but I narrowed it down to specific keys and scenarios, and can reproduce or work around it very reliably.

When I say "upload doesn't work", what I mean is that some physical keys are not updating their behaviour when I upload a new layout that assigns a new character to those keys. Which keys fail seems to depend on their physical position, not on what I'm trying to assign to them. Could be any letter, a special key, or Null – doesn't matter. On the following image the keys that I circled in green are always updating ok, and the keys I crossed out in black keep their old assignment, ignoring the new one (other keys I don't remember). But again, this only concerns the last layer in the layout. If I have three layers in my layout, **all** the keys in the first and second layers will always update fine. It is only the last layer that has this problem.

![Screen Shot 2019-06-28 at 03 06 40](https://user-images.githubusercontent.com/532656/60335313-863ca480-9952-11e9-87ef-3afd50ab8836.png)

So for example, if I have two layers in my layout, and I change the key assigned to one of the crossed out keys on the second layer, and then upload a new layout, that key will not update. If I change one of the keys that is circled green, that circled green key will update. If I upload a layout with multiple keys changed, the green-circle ones will update, but the black-crossed-out-ones will not update.

There are only two ways to work around this that I know of:

1. Add another layer. If you have N layers, all the changes from layers from 1 to N-1 will always upload correctly. So you need to have one extra layer that you won't use because you won't be able to reliably update it.

2. Another option is to upload a layout that has nothing but Nulls in it. When Arbites first opens, it shows a layout with only Nulls in it. If I click "add layer", I get a layout with two layers full of Nulls. If I upload that layout, after that I can upload my real layout that has two layers, it will work – all the keys will be set properly on both layers. But only once. After it's uploaded any further changes will see the same issue with some keys on the second layer not updating. Uploading the nulls layout again and then an updated real layout works though.

Unfortunately, when I upload my real layout after uploading a "null layout" the log is exactly the same as when I upload my real layout and it doesn't work (upload_log_2.txt). So I guess the logs won't be much help here.

To be clear, setting an individual key to Null, uploading that layout, and then setting that key to something else doesn't work – if those keys are problematic ones, they will not be set to null, like any other update this update will be ignored. For whatever reason this works when uploading a whole layout full of nulls though.

---

Hopefully this gives you enough extra information to reproduce this issue. My keyboard is Diverge 3 (left master), Animus 2.7, Arbites 2.3 5000, MacOS 10.14.5.
